### PR TITLE
Deprecate iOS 9

### PR DESCRIPTION
**Pull request by justinswart**
# 📲 What

This deprecates iOS 9 support.

# 🤔 Why

We have a tiny amount of users that are still on iOS 9, currently 0.6% of daily actives. The additional overhead of supporting iOS 9 simply isn't worth it so let's move with the times.

screen shot 2018-11-30 at 2 29 16 pm

# 🛠 How

- Set project target's minimum deployment to iOS 10.3 (latest version of minimum supported major version).
- Remove all `@available` and `#available` checks for iOS 10.0.
- Remove code deprecated in iOS 10.
- Update our notification handling to use `UNUserNotificationCenter`.
- Remove old code that did manual alert dialogues for app store ratings, we get this for free since iOS 10.
- Ensure that notifications are only handled via the `UNUserNotificationCenterDelegate` methods.
  - Notifications are no longer delivered via `application(_:didFinishLaunchingWithOptions:)` so I've removed the related `inputs` from `AppDelegateViewModel` and updated tests. - [reference](https://developer.apple.com/library/archive/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/SchedulingandHandlingLocalNotifications.html#//apple_ref/doc/uid/TP40008194-CH5-SW14).

🚀 screen shot 2018-11-30 at 5 27 16 pm

# 👀 See

Something that's cool that we get for free is this new behaviour for push notifications while the app is open. Instead of the plain old alert dialog we now have the OS push banner:

| < iOS 10  | iOS 10+ |
| ------------- | ------------- |
|  | ezgif com-gif-maker |

# ✅ Acceptance criteria

- [x] Confirm that I haven't deleted code that we still need (happy to pair-review that).
- [x] Can receive and deep-link via remote notifications (app in background).
- [x] Can receive and deep-link via remote notifications (app in foreground).
- [x] Can rate on the app store at the expected time in `ThanksViewController`. This appears to be based on a bunch of criteria defined by the `shouldShowGamesAlert` signal in `ThanksViewModel`.

### CODE REVIEW

File: Kickstarter-iOS/AppDelegate.swift [hunk 1]
```diff
@@ -407,3 +372,24 @@ extension AppDelegate: URLSessionTaskDelegate {
...
+    completionHandler(.alert)
+  }
+
+  public func userNotificationCenter(
+    _: UNUserNotificationCenter,
+    didReceive response: UNNotificationResponse,
+    withCompletionHandler completion: @escaping () -> Void
+    ) {
+    self.viewModel.inputs.didReceive(remoteNotification: response.notification.request.content.userInfo)
+    completion()
```
Comment (justinswart) [inline Kickstarter-iOS/AppDelegate.swift]: I don't think it matters when we call this completion in our case 🤔

File: Kickstarter-iOS/ViewModels/AppDelegateViewModel.swift [hunk 2]
```diff
@@ -280,44 +274,10 @@ AppDelegateViewModelOutputs {
...
-    let remoteNotificationFromLaunch = self.applicationLaunchOptionsProperty.signal.skipNil()
-      .map { _, options in
-        options?[UIApplication.LaunchOptionsKey.remoteNotification] as? [AnyHashable: Any] }
-      .skipNil()
-
-    let localNotificationFromLaunch = self.applicationLaunchOptionsProperty.signal.skipNil()
-      .map { _, options in
-        options?[UIApplication.LaunchOptionsKey.localNotification] as? UILocalNotification }
-      .map { $0?.userInfo }
+    let deepLinkFromNotification = self.remoteNotificationProperty.signal.skipNil()
```
Comment (justinswart) [inline Kickstarter-iOS/ViewModels/AppDelegateViewModel.swift]: This is a lot simpler now that we don't care about whether the app state is active.

File: Kickstarter-iOS/ViewModels/AppDelegateViewModelTests.swift [hunk 3]
```diff
@@ -804,66 +802,12 @@ final class AppDelegateViewModelTests: TestCase {
 
   }
 
-  func testOpenPushNotification_WhileInForeground() {
```
Comment (justinswart) [inline Kickstarter-iOS/ViewModels/AppDelegateViewModelTests.swift]: As far as I could tell we don't really need these tests anymore?

File: Kickstarter-iOS/ViewModels/AppDelegateViewModelTests.swift [hunk 4]
```diff
@@ -1071,26 +985,14 @@ final class AppDelegateViewModelTests: TestCase {
...
+      self.vm.inputs.didReceive(remoteNotification: pushData)
 
       self.goToDashboard.assertValueCount(0)
       self.goToDiscovery.assertValueCount(0)
       self.presentViewController.assertValueCount(0)
     }
   }
 
-  func testOpenNotification_LocalNotification_FromLaunch() {
-    let localNotification = UILocalNotification()
```
Comment (justinswart) [inline Kickstarter-iOS/ViewModels/AppDelegateViewModelTests.swift]: `UILocalNotification` is deprecated.

File: Kickstarter-iOS/Views/Controllers/LiveVideoViewController.swift [hunk 5]
```diff
@@ -148,15 +148,7 @@ public final class LiveVideoViewController: UIViewController {
...
-    do {
-      if #available(iOS 10.0, *) {
-        try AVAudioSession.sharedInstance().setCategory(.playback,
-                                                        mode: .default,
-                                                        options: [])
-      } else {
-        // Apple removed the deprecated method!. Since iOS 12 was release, an idea is to end supporting iOS 9
-      }
-    } catch {}
+    try? AVAudioSession.sharedInstance().setCategory(.playback, mode: .default, options: [])
```
Comment (justinswart) [inline Kickstarter-iOS/Views/Controllers/LiveVideoViewController.swift]: Just doing `try?` here seeing as we weren't handling the error in `catch { }`.

File: Kickstarter-iOS/Views/Controllers/ThanksViewController.swift [hunk 6]
```diff
@@ -214,22 +208,7 @@ internal final class ThanksViewController: UIViewController, UITableViewDelegate
...
-          }, remindHandler: { [weak self] _ in
-            self?.viewModel.inputs.rateRemindLaterButtonTapped()
-          }, noHandler: { [weak self] _ in
-            self?.viewModel.inputs.rateNoThanksButtonTapped()
-        }),
-        animated: true,
-        completion: nil
-      )
-    }
+    SKStoreReviewController.requestReview()
```
Comment (justinswart) [inline Kickstarter-iOS/Views/Controllers/ThanksViewController.swift]: > When you call this method in your shipping app and a rating/review request view is displayed, the system handles the entire process for you.

Apple docs: https://developer.apple.com/documentation/storekit/skstorereviewcontroller/2851536-requestreview

File: Library/PushRegistrationType.swift [hunk 7]
```diff
@@ -4,49 +4,7 @@ import Result
 import UIKit
 import UserNotifications
 
-public enum PushNotificationType: CaseIterable {
```
Comment (justinswart) [inline Library/PushRegistrationType.swift]: Removing this type required no changes at the callsite 🎉

File: Library/ViewModels/ThanksViewModel.swift [hunk 8]
```diff
@@ -167,19 +152,6 @@ public final class ThanksViewModel: ThanksViewModelType, ThanksViewModelInputs,
     self.showGamesNewsletterAlert
       .observeValues { AppEnvironment.current.userDefaults.hasSeenGamesNewsletterPrompt = true }
 
-    project
```
Comment (justinswart) [inline Library/ViewModels/ThanksViewModel.swift]: Seeing as all of this is handled by `SKStoreReviewController.requestReview()` none of this tracking should be needed (nor can it be tracked as far as I can tell).

File: Kickstarter-iOS/ViewModels/AppDelegateViewModelTests.swift (hunk 3 above)
Comment (ifbarrera) [reply Kickstarter-iOS/ViewModels/AppDelegateViewModelTests.swift]: Actually maybe we want to keep one of these as a test for the `Notification Opened` and `Opened Notification` events when the app is in the foreground?

File: Kickstarter-iOS/ViewModels/AppDelegateViewModel.swift [hunk 9]
```diff
@@ -65,17 +65,14 @@ public protocol AppDelegateViewModelInputs {
...
+  func didReceive(remoteNotification notification: [AnyHashable: Any])
 
   /// Call when the app delegate gets notice of a successful notification registration.
   func didRegisterForRemoteNotifications(withDeviceTokenData data: Data)
 
   /// Call when the redirect URL has been found, see `findRedirectUrl` for more information.
   func foundRedirectUrl(_ url: URL)
 
-  /// Call when the user taps "OK" from the notification alert.
-  func openRemoteNotificationTappedOk()
```
Comment (dusi) [inline Kickstarter-iOS/ViewModels/AppDelegateViewModel.swift]: Has been removed from the protocol but still looks like we haven't removed this from `AppDelegateViewModel`

```
fileprivate let openRemoteNotificationTappedOkProperty = MutableProperty(())
public func openRemoteNotificationTappedOk() {
  self.openRemoteNotificationTappedOkProperty.value = ()
}
```

Comment (justinswart) [reply Kickstarter-iOS/ViewModels/AppDelegateViewModel.swift]: Good catch thanks!

File: Kickstarter-iOS/ViewModels/AppDelegateViewModelTests.swift (hunk 3 above)
Comment (justinswart) [reply Kickstarter-iOS/ViewModels/AppDelegateViewModelTests.swift]: Hmm good question - let me take another look to see if it makes sense with the new APIs.

**Comment by justinswart** [comment]
> Do you think we should hold on merging this until 3.10.0 goes out? It might be a good idea to ship this as a separate release to minimize changes on 3.10.0.

@ifbarrera Happy to do that, should be fairly easy to rebase this on that once we've released.

File: Kickstarter-iOS/Views/Controllers/SettingsViewController.swift [hunk 10]
```diff
@@ -110,7 +110,7 @@ final class SettingsViewController: UIViewController {
 
   private func goToAppStore(link: String) {
     guard let url = URL(string: link) else { return }
-    UIApplication.shared.openURL(url)
+    UIApplication.shared.open(url)
```
Comment (dusi) [inline Kickstarter-iOS/Views/Controllers/SettingsViewController.swift]: I've tried searching the code base for `openURL(` and found `UIApplication.shared.openURL(url)` in `Library > SafariActivity.swift:perform()` which now causes compiler warning. Wanna take a look at it?

File: Library/UIAlertController.swift [hunk 11]
```diff
@@ -84,39 +84,6 @@ public extension UIAlertController {
...
-                            remindHandler: @escaping ((UIAlertAction) -> Void),
-                            noHandler: @escaping ((UIAlertAction) -> Void)) -> UIAlertController {
-    let alertController = UIAlertController(
-      title: Strings.profile_settings_rating_title() ,
-      message: Strings.profile_settings_rating_message() ,
-      preferredStyle: .alert
-    )
-    alertController.addAction(
-      UIAlertAction(
-        title: Strings.profile_settings_rating_option_rate_now() ,
```
Comment (dusi) [inline Library/UIAlertController.swift]: Looks like these 3 strings are no longer being used. Wanna get rid of those?

Comment (justinswart) [reply Library/UIAlertController.swift]: Ah thanks, I'll have to open a PR in our strings for that 👍

**Pull request merged**